components/viaduct/android/src/main/java/mozilla/appservices/httpconfig/HttpConfig.kt

```diff
@@ -4,140 +4,108 @@
-                val needed = built.serializedSize
-                val outputBuf = LibViaduct.INSTANCE.viaduct_alloc_bytebuffer(needed)
-                try {
-                    // This is only null if we passed a negative number or something to
-                    // viaduct_alloc_bytebuffer.
-                    val stream = outputBuf.asCodedOutputStream()!!
-                    built.writeTo(stream)
-                    return outputBuf
-                } catch (e: Throwable) {
-                    // Note: we want to clean this up only if we are not returning it to rust.
-                    LibViaduct.INSTANCE.viaduct_destroy_bytebuffer(outputBuf)
-                    LibViaduct.INSTANCE.viaduct_log_error("Failed to write buffer: ${e.message}")
-                    throw e
-                }
-            } finally {
-                LibViaduct.INSTANCE.viaduct_destroy_bytebuffer(b)
-            }
-        }
-    }
-}
-
-internal fun convertMethod(m: MsgTypes.Request.Method): Request.Method {
-    return when (m) {
-        MsgTypes.Request.Method.GET -> Request.Method.GET
-        MsgTypes.Request.Method.POST -> Request.Method.POST
-        MsgTypes.Request.Method.HEAD -> Request.Method.HEAD
-        MsgTypes.Request.Method.OPTIONS -> Request.Method.OPTIONS
-        MsgTypes.Request.Method.DELETE -> Request.Method.DELETE
-        MsgTypes.Request.Method.PUT -> Request.Method.PUT
-        MsgTypes.Request.Method.TRACE -> Request.Method.TRACE
-        MsgTypes.Request.Method.CONNECT -> Request.Method.CONNECT
-    }
-}
-
-internal class CallbackImpl : RawFetchCallback {
+    /**
+     * Important: This can run concurrently on multiple threads
+     *  possibly at the same time. Do not introduce shared memory
+     */
     @Suppress("TooGenericExceptionCaught")
-    override fun invoke(b: RustBuffer.ByValue): RustBuffer.ByValue {
-        try {
-            return RustHttpConfig.doFetch(b)
+    override fun fetch(request: FfiRequest): FfiResponse {
+        // We don't need to lock here because if we end up here,
+        // it means that `setClient` was already called and it
+        // successfully registered this callback.
+        // `setClient` guarantees that the `client` can't change
+        // after it's set initially
+        return try {
+            // The `!!` is safe here because this callback can only be
+            // called after the client is set
+            client!!.value.fetch(request.into()).into(request.method)
         } catch (e: Throwable) {
-            LibViaduct.INSTANCE.viaduct_log_error("doFetch failed: ${e.message}")
-            // This is our last resort. It's bad news should we fail to
-            // return something from this function.
-            return RustBuffer.ByValue()
+            // Ideally, the catching of the error should happen in `uniffi`
+            // generated code and it should be converted to a proper Rust Error
+            // but until we add support for that, this should do
```

Comment:
I don't quite understand this comment - who needs to add support, and for what exactly? :) Also wondering if the `.into()`s should be outside the exception handler?

Reply:
Good catch, the comment isn't clear at all 😅 I'll add a reference to https://github.com/mozilla/uniffi-rs/issues/351 and make it clearer!